policies/recipes/recipedb.go

```diff
@@ -15,32 +15,85 @@
 package recipes
 
 import (
+	"encoding/json"
 	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
 	"strconv"
 	"strings"
+	"time"
+)
+
+var (
+	dbPath   = "/var/lib/google/osconfig_recipedb"
+	tempPath = filepath.Join(os.TempDir(), "osconfig_recipedb_temp")
 )
 
 // RecipeDB represents local state of installed recipes.
-type RecipeDB struct{}
+type RecipeDB struct {
+	recipes map[string]Recipe
+}
 
-func newRecipeDB() RecipeDB {
-	return RecipeDB{}
+func newRecipeDB() (*RecipeDB, error) {
+	f, err := os.Open(dbPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return &RecipeDB{recipes: make(map[string]Recipe)}, nil
+		}
+		return nil, err
+	}
+	defer f.Close()
+	bytes, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	db := &RecipeDB{}
+	if err := json.Unmarshal(bytes, &db); err != nil {
+		return nil, err
+	}
+	return db, nil
 }
 
 // GetRecipe returns the Recipe object for the given recipe name.
 func (db *RecipeDB) GetRecipe(name string) (Recipe, bool) {
-	return Recipe{}, false
+	r, ok := db.recipes[name]
+	return r, ok
 }
 
 // AddRecipe marks a recipe as installed.
 func (db *RecipeDB) AddRecipe(name, version string) error {
+	versionNum, err := convertVersion(version)
+	if err != nil {
+		return err
+	}
+	db.recipes[name] = Recipe{name: name, version: versionNum, installTime: time.Now().Unix()}
+	dbBytes, err := json.Marshal(db)
+	if err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Dir(dbPath), 0755); err != nil {
+		return err
+	}
+
+	f, err := os.Create(tempPath)
```

Comment:
The tempPath should be the same as the dbPath with just a temp string appended. Moves across file systems are not atomic.

Reply:
Okay, I put it in the same directory and am using ioutil.TempFile to create the file and choose the exact name.